### DISCUSSION TITLE
Deleting CDF zVar with pycdf seems to not fully work

### DISCUSSION THREAD

**Discussion by xbonnin** [spacepy/spacepy#504] [Support and Q&A] [answered]
Hello, 

I would like to delete several zVars from a large CDF file (~10 GB) using pycdf.
I used the `del cdf[zvar]` command (as described in https://spacepy.github.io/autosummary/spacepy.pycdf.Var.html#spacepy.pycdf.Var)

It seems to work because the zVariables do not appear anymore in the CDF after the operation.
However the data size of the file remains the same (~10 GB).

For testing, I also tried to apply `del cdf[zvar]` command  over the whole CDF file, in order to delete all the zVariables.
After the operation, all the zVariables are disappeared and the CDF looks empty, but the data size of the file is the same again (~10 GB).

It's like if all the zVariable class descriptions are well erased from the CDF, but not the data it-self.

Do I miss something? 

Thanks for your help.

### Minimal example to reproduce issue:

Here is below a piece of code I used to test the pycdf behaviour w.r.t. zVariable deletion.
Running this code allows to remove zVariables from CDF, but the size of the file remains the same before/after.

```
from pathlib import Path
from spacepy.pycdf import CDF

cdf_file = Path('solo_L2_rpw-tds-surv-rswf-e-cdag_20201227_V04.cdf')

print(f'Opening {cdf_file} ...')
cdf = CDF(str(cdf_file))
cdf.readonly(False)

zvar_names = cdf.keys()

if len(zvar_names) > 0:
    for zvar in zvar_names:
        print(f'Deleting {zvar} zVariable ...')
        # del cdf[zvar][...] # attempt to remove all data before deleting the zVar from CDF
        del cdf[zvar]

cdf.close()
```

### Error message/Traceback:
No error message returned

### OS, Python version, and dependency version information:
The operation was run in the two following environments (with the same result):

macOS-10.14.6-x86_64-i386-64bit
sys.version_info(major=3, minor=8, micro=6, releaselevel='final', serial=0)
numpy=1.19.5
scipy=1.6.0
matplotlib=3.3.3

and

Linux-4.19.121-linuxkit-x86_64-with-glibc2.2.5
sys.version_info(major=3, minor=8, micro=7, releaselevel='final', serial=0)
numpy=1.19.5
scipy=1.5.4
matplotlib=3.3.4

### Version of SpacePy
I used Spacepy V0.2.2 and CDF V3.7.1

**Comment by jtniehof** [comment, marked as answer]
This is the way CDF works. Deleting a variable makes the space available within the file but does not reduce the file size. See page 49 of the [CDF User's Guide](https://spdf.gsfc.nasa.gov/pub/software/cdf/doc/cdf380/cdf380ug.pdf) (2.3.7, "Deleting"), also the "Allocated Records" information on pg. 55 (under 2.3.12, "Records"). Even using `cdfconvert` doesn't usually recover the space.

If you need to shrink a file, the most reliable way is to create a new CDF and use [clone](https://spacepy.github.io/autosummary/spacepy.pycdf.CDF.html#spacepy.pycdf.CDF.clone) to copy the desired variables over:

```python
import spacepy.pycdf

desired_variables = ['Epoch', 'Flux']
with spacepy.pycdf.CDF('old_cdf.cdf') as oldfile:
    with spacepy.pycdf.CDF('new_cdf.cdf', create=True) as newfile:
        newfile.attrs.clone(oldfile.attrs)
        for v in desired_variables:
            newfile.clone(oldfile[v])
```

If you don't want to keep manual track of dependencies, methods of [VarBundle](https://spacepy.github.io/autosummary/spacepy.pycdf.istp.VarBundle.html#spacepy.pycdf.istp.VarBundle) will take care of that for you:
```python
import spacepy.pycdf
import spacepy.pycdf.istp

desired_variables = ['Flux'] # Dependencies of Flux will come automatically
with spacepy.pycdf.CDF('old_cdf.cdf') as oldfile:
    with spacepy.pycdf.CDF('new_cdf.cdf', create=True) as newfile:
        newfile.attrs.clone(oldfile.attrs)
        for v in desired_variables:
            spacepy.istp.VarBundle(oldfile[v]).output(newfile)
```

CDF file properties like majority and whole-file compression would have to be set manually, e.g. `newfile.col_major(oldfile.col_major)`. `clone` will copy the per-variable compression (and type, variance and such), as will the `output` method of `VarBundle`.

**Comment by xbonnin** [comment]
Thanks for the quick reply.
I was not awared of this CDF feature (I should read the F... manual :-)).

I've tested your solution and it works! 
(I just need to add `pycdf.lib.set_backward(False)` line to avoid the issue defined in https://github.com/spacepy/spacepy/issues/57). 

Just a question: if I also want to clone the CDF file properties (i.e., CDF majority, compression, etc.), do I need to add commands manually  (i.e., `newfile.col_major(oldfile.col_major)`, etc.)

**Reply by jtniehof** [reply]
The manual's big...it's not always going to be easy to find exactly what you want in it, so no problem asking.

The backward-compatibility default will change in 0.3 so that'll be one less thing you have to do in the future.

I've converted this issue to a Q&A and done some light editing of the conversation: moving some of my comments into this reply, others into my answer above, and also copying some of your comments in there so it's closer to a one-stop answer if others have this question.

**Comment by xbonnin** [comment]
Ok thanks again!
